dsl/OpusEnum.cc

```diff
@@ -38,34 +40,104 @@ bool isOpusEnum(core::MutableContext ctx, ast::ClassDef *klass) {
     return id->symbol == core::Symbols::root();
 }
 
-void OpusEnum::patchDSL(core::MutableContext ctx, ast::ClassDef *klass) {
-    if (ctx.state.runningUnderAutogen) {
-        // TODO(jez) Verify whether this DSL pass is safe to run in for autogen
-        return;
+vector<unique_ptr<ast::Expression>> processStat(core::MutableContext ctx, ast::ClassDef *klass, ast::Expression *stat) {
+    auto *asgn = ast::cast_tree<ast::Assign>(stat);
+    if (asgn == nullptr) {
+        return {};
     }
 
-    if (!isOpusEnum(ctx, klass)) {
-        return;
+    auto *lhs = ast::cast_tree<ast::UnresolvedConstantLit>(asgn->lhs.get());
+    if (lhs == nullptr) {
+        return {};
+    }
+
+    auto *rhs = ast::cast_tree<ast::Send>(asgn->rhs.get());
+    if (rhs == nullptr) {
+        return {};
+    }
+
+    if (rhs->fun != core::Names::new_() && rhs->fun != core::Names::let()) {
+        return {};
+    }
+
+    if (rhs->fun == core::Names::new_() && !rhs->recv->isSelfReference()) {
+        return {};
     }
 
-    for (auto &stat : klass->rhs) {
-        auto *asgn = ast::cast_tree<ast::Assign>(stat.get());
-        if (asgn == nullptr) {
-            continue;
+    if (rhs->fun == core::Names::let()) {
+        auto recv = ast::cast_tree<ast::UnresolvedConstantLit>(rhs->recv.get());
+        if (recv == nullptr) {
+            return {};
         }
 
-        auto *rhs = ast::cast_tree<ast::Send>(asgn->rhs.get());
-        if (rhs == nullptr) {
-            continue;
+        if (rhs->args.size() != 2) {
+            return {};
         }
 
-        if (!(rhs->fun == core::Names::new_() && rhs->recv->isSelfReference())) {
-            continue;
+        auto arg0 = ast::cast_tree<ast::Send>(rhs->args[0].get());
+        if (arg0 == nullptr) {
+            return {};
         }
 
-        auto loc = asgn->rhs->loc;
-        asgn->rhs = ast::MK::Send2(loc, ast::MK::T(loc), core::Names::let(), move(asgn->rhs), ast::MK::Self(loc));
+        if (!(arg0->fun == core::Names::new_() && arg0->recv->isSelfReference())) {
+            return {};
+        }
     }
+
+    auto name = ctx.state.freshNameUnique(core::UniqueNameKind::OpusEnum, lhs->cnst, 1);
+    auto classCnst = ast::MK::UnresolvedConstant(lhs->loc, ast::MK::EmptyTree(), name);
+    ast::ClassDef::ANCESTORS_store parent;
+    parent.emplace_back(klass->name->deepCopy());
+    ast::ClassDef::RHS_store classRhs;
+    classRhs.emplace_back(ast::MK::Send1(stat->loc, ast::MK::Self(stat->loc), core::Names::extend(),
+                                         ast::MK::Constant(stat->loc, core::Symbols::T_Helpers())));
+    classRhs.emplace_back(ast::MK::Send1(stat->loc, ast::MK::Self(stat->loc), core::Names::include(),
+                                         ast::MK::Constant(stat->loc, core::Symbols::Singleton())));
+    classRhs.emplace_back(ast::MK::Send0(stat->loc, ast::MK::Self(stat->loc), core::Names::declareFinal()));
```

Comment:
Will it be a problem that it introduces these three for every assignment? Would it be better to set a flag, and add them in `postTransformClassDef`?

Reply:
Never mind, this isn't going onto the class being examined, but the one being generated.